### PR TITLE
Fix `CosineAnnealingLR` on restart

### LINKED ISSUE #60265: LR Scheduler bug when resuming training from checkpoint
## 🐛 Bug

When trying to resume training from a given checkpoint using the CosineAnnealingLR scheduler, there's a discontinuity in learning rate values that I can't explain nor fix!

Let’s say I trained a model for 45 epochs, then stopped training and saved both the optimizer state and the scheduler state.
I want to resume training from epoch 46, until epoch 100. To do that, I simply need to instantiate a new LR scheduler with last_epoch = 45. But when looking at learning rates values as a function of epochs, I get a discontinuity at epoch 45.

## To Reproduce

```
model = torch.nn.Sequential(
            OrderedDict(
                [
                    ('conv1', torch.nn.Conv2d(3, 20, 3)),
                    ('relu1', torch.nn.ReLU()),
                    ('conv2', torch.nn.Conv2d(20, 64, 3)),
                    ('relu2', torch.nn.ReLU()),
                ]
            )
        )

base_lr = 0.001

epoch_start = 0
epoch_stop = 45
epoch_end = 100

optimizer = torch.optim.Adam(model.parameters(), lr=base_lr)
scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=epoch_end, last_epoch=epoch_start-1)

# run 45 epochs and save learning rate values
lr_values = []
for i in range(epoch_start, epoch_stop):
    lr = scheduler.get_last_lr()[0]
    lr_values.append(lr)
    optimizer.step()
    scheduler.step()

# stop training and save both optimizer and scheduler states
optimizer_state = optimizer.state_dict()
scheduler_state = scheduler.state_dict()

# resume training 
epoch_restart = scheduler_state['last_epoch']
new_optimizer = torch.optim.Adam(model.parameters(), lr=base_lr)
new_optimizer.load_state_dict(optimizer_state)
new_scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(new_optimizer, T_max=epoch_end, last_epoch=epoch_restart-1)

# run remaining epochs = []
for i in range(epoch_restart, epoch_end):
    lr = new_scheduler.get_last_lr()[-1]
    lr_values.append(lr)
    new_optimizer.step()
    new_scheduler.step()

# plot learning rate values as a function of epochs
plt.plot(range(epoch_start, epoch_end), lr_values)
plt.xlabel('epoch')
plt.ylabel('lr')
plt.title('resume training from checkpoint')
plt.show()
```

<a href="https://imgbb.com/">resume-from-checkpoint</a>

You can notice a discontinuity at epoch 45

## Expected behavior

Let's run uninterrupted 100 epochs long training to see what would be the expected learning values

```
model = torch.nn.Sequential(
            OrderedDict(
                [
                    ('conv1', torch.nn.Conv2d(3, 20, 3)),
                    ('relu1', torch.nn.ReLU()),
                    ('conv2', torch.nn.Conv2d(20, 64, 3)),
                    ('relu2', torch.nn.ReLU()),
                ]
            )
        )

base_lr = 0.001

epoch_start = 0
epoch_end = 100

optimizer = torch.optim.Adam(model.parameters(), lr=base_lr)
scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=epoch_end, last_epoch=epoch_start-1)

# run 100 epochs and save learning rate values
lr_values = []
for i in range(epoch_start, epoch_end):
    lr = scheduler.get_last_lr()[0]
    lr_values.append(lr)
    optimizer.step()
    scheduler.step()

# plot learning rate values as a function of epochs
plt.plot(range(epoch_start, epoch_end), lr_values)
plt.xlabel('epoch')
plt.ylabel('lr')
plt.title('one way training')
plt.show()
```

<a href="https://imgbb.com/">one-way-training</a>

## Environment

 - PyTorch Version: 1.5.1
 - Python version: Python 3.6.9

cc @vincentqb @jbschlosser @albanD @iramazanli

### PR DESCRIPTION
**Pull request by emcastillo**
Fixes #60265

The initial LR for this scheduler is not consistent when a new instance is created with `last_epoch != -1`

Maybe we can refactor the testing code to test `last_epoch != -1` in schedulers that can recreate their state from the current epoch?

### CODE REVIEW

**Review by albanD — APPROVED** [review]
The change looks good!
Could you rebase on top of master to make sure latest CI run please?

**Comment by emcastillo** [comment]
rebased!
